File: solution/576. Out of Boundary Paths/576.cpp
class Solution {
public:
    int findPaths(int m, int n, int maxMove, int startRow, int startColumn) {
        const int Mod = 1e9 + 7;
        vector<vector<int>> dp(m, vector<int>(n));
        auto temp = dp;
        for (int k = 0; k < maxMove; ++k) {
            for (int i = 0; i < m; ++i) {
                for (int j = 0; j < n; ++j) {
                    long up = i - 1 < 0 ? 1 : dp[i - 1][j];
                    long down = i + 1 >= m ? 1 : dp[i + 1][j];
                    long left = j - 1 < 0 ? 1 : dp[i][j - 1];
                    long right = j + 1 >= n ? 1 : dp[i][j + 1];
                    temp[i][j] = (up + down + left + right) % Mod;
                }
            }
            swap(temp, dp);
        }
        return dp[startRow][startColumn];
    }
};

File: solution/576. Out of Boundary Paths/576.py
class Solution:
    def findPaths(self, m: int, n: int, maxMove: int, startRow: int, startColumn: int) -> int:
        MOD = 10**9 + 7
        dp = [[0] * (n + 2) for _ in range(m + 2)]
        
        for i in range(1, m + 1):
            dp[i][0] = dp[i][n + 1] = 1
        for j in range(1, n + 1):
            dp[0][j] = dp[m + 1][j] = 1
        
        pre = [row[:] for row in dp]
        
        for k in range(maxMove):
            for i in range(1, m + 1):
                for j in range(1, n + 1):
                    dp[i][j] = (pre[i - 1][j] + pre[i + 1][j] + pre[i][j - 1] + pre[i][j + 1]) % MOD
            pre, dp = dp, pre
            
        return pre[startRow + 1][startColumn + 1]
    
